### PR TITLE
Support v3 APIs and basic auth in Kafka Contact Point

### PR DESCRIPTION
**Pull request by ghost**
What?
---
This commit does the following,
- Adds basic auth to the Kafka Contact Point.
- Adds support for v3 Kafka APIs

Basic Auth
---
When creating a Kafka Rest Proxy Contact Point on Grafana, users should now be able to use basic auth.
They can specific a username and can choose to either specify the password or let the password be read from a file on
the grafana server.
Like this,

Screenshot 2023-01-08 at 11 34 10 PM

Additionally, when the user chooses to supply the password as a file on the server, the password is re-read from the file on
an API failure. This is to allow the password to be changed without needing to restart the grafana server.

v3 Kafka APIs
----
The existing Kafka Contact Point only supported v2 APIs. While this was fine till now - newer systems are moving towards v3. Confluent is already only deploying v3 APIs for their Kafka Cloud Clusters.
This commit adds support to allow using v3 APIs to publish messages to kafka.

When creating the contact point, the user can now choose the API version they would like to use. By default, v2 will be used.

Screenshot 2023-01-08 at 11 41 19 PM

Testing
----
1. Added tests in kafka_tests.go
2. Manually tested this against my grafana deployment. Existing v2 APIs continued to work fine. Tested this using the "Test Contact Point" in the "Edit Contact Point" page.
3. Manually tested that I was able to messages to Kafka v3
4. [Done] ~TODO: Verify that the contact points created before this change continue to load fine after upgrading to this.~ Tested upgrade from 9.3.1 to main(v9.4.0-pre (d9be87c22d)) The previously provisioned kafka contact point was working as expected after the upgrade.

Things I need help with
-----
~1. How do I build and test this with Grafana?~
2. How do I know/specify which Grafana's release will this go into?

CHANGELOG
-----
- Kafka Contact Point now supports basic auth.
- Kafka Contact Point now supports v3 APIs.

References
----
1. [Kafka v2/v3 API comparison doc](https://docs.confluent.io/platform/current/kafka-rest/api.html)
2. Corresponding changes in Grafana UI: https://github.com/grafana/grafana/pull/61123

Credits
------
Something similar was done in this PR. Which I referenced: https://github.com/grafana/grafana/pull/41227

### CODE REVIEW

**Comment by ghost** [comment]
Hey Guys!
Do I need a corresponding change in [available_channels.go](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/notifier/channels_config/available_channels.go) as well?

**Comment by ghost** [comment]
Related issue: https://github.com/grafana/grafana/issues/41176

**Comment by alexweav** [comment]
Hi, thanks very much for this contribution :wave:

At first glance, this seems like a reasonable change!

> Do I need a corresponding change in [available_channels.go](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/notifier/channels_config/available_channels.go) as well?

Yes, this will automatically create UI elements to match your new fields, so they will be visible and editable in the UI.

Regarding password in the DB vs. in a file, there are a few additional considerations to take into account:
- Grafana [encrypts](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-database-encryption/) secrets stored in the database.
- Network-based provisioning systems (Terraform, etc) do not have access to the disk of the instance, and will not be able to create a password file.
- Many users expect all of Grafana's data to be in the database, and many Grafana deployments have totally volatile filesystems.
- In Grafana's High-Availability mode, a password file would have to be synchronized among all instances, where a password in the database only has to be set once.

Additionally, some pointers that may be of help here:

> a password changes shouldn't require the user to restart the grafana server

We offer a way around this, sending a `POST` request to `/api/admin/provisioning/alerting/reload` should [reload and re-sync](https://grafana.com/docs/grafana/latest/developers/http_api/admin/#reload-provisioning-configurations) the provisioning files without a server restart.

> I would prefer to store the kafka password in a kubernates secret and mount it onto the Grafana pod as a file.

The provisioning files also allow environment variables to be referenced. A common pattern that we see is to mount the secret as an environment variable rather than a file, then [have the provisioning file point to it.](https://grafana.com/docs/grafana/latest/administration/provisioning/#using-environment-variables)

**Comment by Ghazanfar-CFLT** [comment]
Thanks for the review @alexweav !
>Yes, this will automatically create UI elements to match your new fields, so they will be visible and editable in the UI.

Sounds good, I will raise another PR for this.

I think I agree with your points of needing to allow for a password.
Would it be okay if I allowed both options? Either supply the password or supply a path to a file containing the password?

>The provisioning files also allow environment variables to be referenced. A common pattern that we see is to mount the secret as an environment variable rather than a file, then [have the provisioning file point to it](https://grafana.com/docs/grafana/latest/administration/provisioning/#using-environment-variables)

The problem with this is that changes in the secret will not propagate to the container's ENV variable.
From the [Kubernetes doc](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables),
>Note: If a container already consumes a Secret in an environment variable, a Secret update will not be seen by the container unless it is restarted. There are third party solutions for triggering restarts when secrets change.

This is where I am coming from: My workplace wants us rotating passwords very frequently - every 15 mins or so. So, I would prefer the password be in a secret and mounted as a file. I would like to avoid frequent restarts of Grafana.

Given that there are uses for both the approaches, shall I go ahead with supporting them both please?

**Comment by ghost** [comment]
Thanks for the review @alexweav !
>Yes, this will automatically create UI elements to match your new fields, so they will be visible and editable in the UI.

Sounds good, I will raise another PR for this.

I think I agree with your points of needing to allow for a password.
Would it be okay if I allowed both options? Either supply the password or supply a path to a file containing the password?

>The provisioning files also allow environment variables to be referenced. A common pattern that we see is to mount the secret as an environment variable rather than a file, then [have the provisioning file point to it](https://grafana.com/docs/grafana/latest/administration/provisioning/#using-environment-variables)

The problem with this is that changes in the secret will not propagate to the container's ENV variable.
From the [Kubernetes doc](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables),
>Note: If a container already consumes a Secret in an environment variable, a Secret update will not be seen by the container unless it is restarted. There are third party solutions for triggering restarts when secrets change.

This is where I am coming from: My workplace wants us rotating passwords very frequently - every 15 mins or so. So, I would prefer the password be in a secret and mounted as a file. I would like to avoid frequent restarts of Grafana.

Given that there are uses for both the approaches, shall I go ahead with supporting them both please?

**Comment by ghost** [comment]
@alexweav Here are the corresponding changes for the UI: https://github.com/grafana/grafana/pull/61123

**Comment by ghost** [comment]
Given https://github.com/grafana/grafana/pull/60655, I am assuming that the earliest release this change can only go into is 9.4. Is that correct?

**Comment by ghost** [comment]
Added upgrade test results to the PR description.

File: alerting/notifier/channels/kafka.go
Comment (yuri-tceretian) [inline alerting/notifier/channels/kafka.go]: I do not think that we are ready to support this in Grafana notifiers, at least in the current implementation.
1. I think this creates quite a serious security vulnerability because there is nothing to stop a malicious user that has access to API to specify a path to SQLlite database, or grafana config file and send it in the Authz header back to malicious API endpoint.
2. None of Grafana notifiers support loading secrets from file. This might change when we start replacing Grafana implementation of notifiers with the Alertmanager ones, and we will have to figure out how to make it secure.

cc @gotjosh what do you think?

File: alerting/notifier/channels/kafka.go
Comment (yuri-tceretian) [inline alerting/notifier/channels/kafka.go]: Let's rename it to Password to match the name of the json field.

File: alerting/notifier/channels/kafka.go
Comment (yuri-tceretian) [inline alerting/notifier/channels/kafka.go]: I think we do not need this field. I think validating that both user and password are specified would be enough

File: alerting/notifier/channels/kafka.go
Comment (yuri-tceretian) [inline alerting/notifier/channels/kafka.go:192]: can you add check here
```
if tmplErr != nil {
		kn.log.Warn("failed to template Kafka URL", "error", tmplErr.Error())
	}
```

**Review by yuri-tceretian — COMMENTED** [review]
Generally, good change. Only serious concern is about loading secrets from file.

File: alerting/notifier/channels/kafka.go
Comment (gotjosh) [reply alerting/notifier/channels/kafka.go]: Agreed with Yuri here, this does not belong within Grafana. We already have a way in Mimir to unsupport these fields in Cloud. 

So when we move over, we'll get it for free.

Comment (ghost) [reply alerting/notifier/channels/kafka.go]: Got it! That's a valid concern.
I would also like to point out that it is current supported to supply Grafana's DB password from a file here: https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#file-provider

Would it help if I limit the size of the password on the file? To say 32 characters? 

The reasons I wanted to support this is so that when Grafana is deployed in a kubernetes environment, I would prefer that the password is stores in a kubernetes secret and mounted onto the grafana container. This also allows me to rotate the password frequently without restarting grafana each time.

Comment (yuri-tceretian) [reply alerting/notifier/channels/kafka.go]: The difference between these two cases is that Grafana config can only be changed via updating a config file or via environment variable, and that requires access to the host. In the case of notification policies, this is a purely API call, which has much more relaxed constraints. Basically, any user with `Editor` role is able to create a contact point.

> This also allows me to rotate the password frequently without restarting grafana each time.

You can use an API call to rotate secrets instead. It can be a sidecar container that would do that.

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go:192]: Sorry, there was a check here, I added additional code in the middle and it got moved down to line 235. I'll fix this!

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go]: Would it help if I limit the size of the password on the file? To say 32 characters? Or would you guys prefer that I remove this altogether?

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go]: Sounds good. I'll remove this.

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go]: Agreed, I'll make the change!

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go]: Thanks @yuri-tceretian 
I'll remove this part and leave you guys an update. Thanks for the review!

Comment (yuri-tceretian) [reply alerting/notifier/channels/kafka.go]: One way I can recommend: introduce support for kafka in the alertmanager repository https://github.com/prometheus/alertmanager where you can introduce support for loading secrets from files. Then if it is accepted, you can configure Grafana to route all alerts to external Alertmanager instead of the internal one.

I believe opensource community would thank you for that!

Comment (ghost) [reply alerting/notifier/channels/kafka.go]: Ah, that makes sense! Thanks @yuri-tceretian! I'll see if I can find the time to work on that.

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go:192]: Done!

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go]: I have removed validation for these altogether. Give both, give either, don't give anything, I just pass it along. Does that work @yuri-tceretian ? Or would you prefer that I validate that if one is provided, the other should be too?

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go]: Done!

File: alerting/notifier/channels/kafka.go
Comment (yuri-tceretian) [inline alerting/notifier/channels/kafka.go]: can it be Username and JSON\yaml `username`?

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go]: Have remove the reading from file part. Will mark this resolved. I will get to the alertmanager changes in the future.

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go]: On it!

Comment (ghost) [reply alerting/notifier/channels/kafka.go]: Done!

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go]: From here: https://stackoverflow.com/a/74658311/4042839
it looks like username and/or password can be empty. So, I don't think we should validate that if one is provided, the other should be too.

@yuri-tceretian does that answer it? Can I mark this as resolved?

Comment (yuri-tceretian) [reply alerting/notifier/channels/kafka.go]: Although I think this is not a good practice, I am ok.

File: alerting/notifier/channels/kafka.go
Comment (yuri-tceretian) [inline alerting/notifier/channels/kafka.go:95]: Password is supposed to be a secret. So please add 
```suggestion
	settings.Password = fc.DecryptFunc(context.Background(), fc.Config.SecureSettings, "password", settings.Password)
```

**Comment by yuri-tceretian** [comment]
LGTM after the suggestion above is committed.

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go]: @yuri-tceretian I can add this if you prefer :)
But won't that be adding limitations on top of features allowed by Basic Auth? If Basic Auth allows username or password to be empty, then we should allow it as well, right?

File: alerting/notifier/channels/kafka.go
Comment (ghost) [reply alerting/notifier/channels/kafka.go:95]: Added!

**Review by yuri-tceretian — APPROVED** [review]
LGTM. Thanks!

**Comment by ghost** [comment]
Thank you for reviewing @yuri-tceretian !
Am I good to merge this?

**Comment by ghost** [comment]
Oh, I can't merge it! It says "Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests."

**Comment by ghost** [comment]
Can someone also look at the corresponding UI changers on grafana please? https://github.com/grafana/grafana/pull/61123

**Pull request merged**
